signer/src/utxo.rs

```diff
@@ -0,0 +1,976 @@
+use bitcoin::absolute::LockTime;
+use bitcoin::key::Secp256k1;
+use bitcoin::secp256k1::SECP256K1;
+use bitcoin::taproot::LeafVersion;
+use bitcoin::taproot::NodeInfo;
+use bitcoin::taproot::Signature;
+use bitcoin::taproot::TaprootSpendInfo;
+use bitcoin::transaction::Version;
+use bitcoin::Address;
+use bitcoin::Amount;
+use bitcoin::OutPoint;
+use bitcoin::PublicKey;
+use bitcoin::ScriptBuf;
+use bitcoin::Sequence;
+use bitcoin::Transaction;
+use bitcoin::TxIn;
+use bitcoin::TxOut;
+use bitcoin::Witness;
+use bitcoin::XOnlyPublicKey;
+use secp256k1::Keypair;
+use secp256k1::Message;
+
+use crate::error::Error;
+use crate::packaging::compute_optimal_packages;
+use crate::packaging::Weighted;
+
+#[derive(Debug, Clone, Copy)]
+pub struct SignerBtcState {
+    /// The outstanding signer UTXO.
+    pub utxo: SignerUtxo,
+    /// The current market fee rate in sat/vByte.
+    pub fee_rate: u64,
+    /// The current public key of the signers
+    pub public_key: PublicKey,
+}
+
+#[derive(Debug)]
+pub struct SbtcRequests {
+    /// Accepted and pending deposit requests.
+    pub deposits: Vec<DepositRequest>,
+    /// Accepted and pending withdrawal requests.
+    pub withdrawals: Vec<WithdrawalRequest>,
+    /// Summary of the Signers' UTXO and information necessary for
+    /// constructing their next UTXO.
+    pub signer_state: SignerBtcState,
+    /// The minimum acceptable number of votes for any given request.
+    pub accept_threshold: u32,
+    /// The total number of signers.
+    pub num_signers: u32,
+}
+
+impl SbtcRequests {
+    /// Construct the next transaction package given requests and the
+    /// signers' UTXO.
+    ///
+    /// This function can fail if the output amounts are greater than the
+    /// input amounts.
+    pub fn construct_transactions(&self) -> Result<Vec<UnsignedTransaction>, Error> {
+        if self.deposits.is_empty() && self.withdrawals.is_empty() {
+            tracing::info!("No deposits or withdrawals so no BTC transaction");
+            return Ok(Vec::new());
+        }
+
+        let withdrawals = self.withdrawals.iter().map(Request::Withdrawal);
+        let deposits = self.deposits.iter().map(Request::Deposit);
+
+        // Create a list of requests where each request can be approved on its own.
+        let items = deposits.chain(withdrawals);
+
+        compute_optimal_packages(items, self.reject_capacity())
```

Comment:
I'm a bit confused as to where input/output weight come into play here. In general, inputs will be very similar size (though not exact), and outputs will be varying in size (depending on the type of address). So, wouldn't this function need to have some mechanism of knowing a request's "cost"?

Reply:
Good question. The `compute_optimal_packages` function computes approval sets using only how the signers have voted on each of the requests. The number of rejections is the request's "weight", it is completely separate from input/output weight of a BTC transaction. That function doesn't care about the BTC weight at all.

You know, I should rename that term but I couldn't think of anything better (size? heaviness? pounds? lbs?). Everything I thought of kinda missed the mark a little but maybe weight is even worse.